File: wix-restaurants-java-client/src/main/java/com/wix/restaurants/DefaultWixRestaurantsClient.java
package com.wix.restaurants;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.openrest.v1_1.*;
import com.openrest.v1_1.Error;
import com.wix.restaurants.authentication.DefaultWixRestaurantsAuthenticationClient;
import com.wix.restaurants.authentication.WixRestaurantsAuthenticationClient;
import com.wix.restaurants.authorization.Role;
import com.wix.restaurants.authorization.requests.GetRoleRequest;
import com.wix.restaurants.exceptions.*;
import com.wix.restaurants.i18n.Locale;
import com.wix.restaurants.orders.Statuses;
import com.wix.restaurants.orders.requests.*;
import com.wix.restaurants.requests.*;
import com.wix.restaurants.reservations.Reservation;
import com.wix.restaurants.reservations.ReservationsResponse;
import com.wix.restaurants.reservations.requests.*;

import java.io.IOException;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Map;

public class DefaultWixRestaurantsClient implements WixRestaurantsClient {
    private final OpenrestClient openrest;
    private final WixRestaurantsAuthenticationClient authenticationClient;

    public static class Builder {
        private HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
        private int connectTimeout = 0;
        private int readTimeout = 0;
        private int numberOfRetries = 0;
        private String authApiUrl = com.wix.restaurants.authentication.Endpoints.PRODUCTION;
        private String apiUrl = Endpoints.production;

        public Builder setRequestFactory(HttpRequestFactory requestFactory) {
            this.requestFactory = requestFactory;
            return this;
        }

        public Builder setConnectTimeout(int connectTimeout) {
            this.connectTimeout = connectTimeout;
            return this;
        }

        public Builder setReadTimeout(int readTimeout) {
            this.readTimeout = readTimeout;
            return this;
        }

        public Builder setNumberOfRetries(int numberOfRetries) {
            this.numberOfRetries = numberOfRetries;
            return this;
        }

        public Builder setAuthApiUrl(String authApiUrl) {
            this.authApiUrl = authApiUrl;
            return this;
        }

        public Builder setApiUrl(String apiUrl) {
            this.apiUrl = apiUrl;
            return this;
        }

        public DefaultWixRestaurantsClient build() {
            return new DefaultWixRestaurantsClient(
                    requestFactory, connectTimeout, readTimeout, numberOfRetries,
                    authApiUrl, apiUrl);
        }
    }

    public DefaultWixRestaurantsClient(HttpRequestFactory requestFactory,
                                       Integer connectTimeout, Integer readTimeout, Integer numberOfRetries,
                                       String authApiUrl, String apiUrl) {
        openrest = new OpenrestClient(requestFactory, connectTimeout, readTimeout, numberOfRetries, apiUrl);
        authenticationClient = new DefaultWixRestaurantsAuthenticationClient(
                requestFactory, connectTimeout, readTimeout, numberOfRetries, authApiUrl);
    }

    public DefaultWixRestaurantsClient(HttpRequestFactory requestFactory,
                                       Integer connectTimeout, Integer readTimeout, Integer numberOfRetries) {
        this(requestFactory, connectTimeout, readTimeout, numberOfRetries,
                com.wix.restaurants.authentication.Endpoints.PRODUCTION,
                Endpoints.production);
    }

    @Override
    public WixRestaurantsAuthenticationClient getAuthenticationClient() {
        return authenticationClient;
    }

    @Override
    public RestaurantFullInfo retrieveRestaurantInfo(String restaurantId) {
        final GetOrganizationFullRequest getOrganizationFullRequest = new GetOrganizationFullRequest();
        getOrganizationFullRequest.organizationId = restaurantId;

        final RestaurantFullInfo getOrganizationFullResponse = request(
                getOrganizationFullRequest, new TypeReference<Response<RestaurantFullInfo>>() {});

        return getOrganizationFullResponse;
    }

    @Override
    public Organization retrieveOrganization(String organizationId) {
        final GetOrganizationRequest getOrganizationRequest = new GetOrganizationRequest();
        getOrganizationRequest.organizationId = organizationId;

        final Organization getOrganizationResponse = request(
                getOrganizationRequest, new TypeReference<Response<Organization>>() {});

        return getOrganizationResponse;
    }

    @Override
    public Organization setOrganization(String accessToken, Organization organization) {
        final SetOrganizationRequest setOrganizationRequest = new SetOrganizationRequest();
        setOrganizationRequest.accessToken = accessToken;
        setOrganizationRequest.organization = organization;

        final Organization setOrganizationResponse = request(
                setOrganizationRequest, new TypeReference<Response<Organization>>() {});

        return setOrganizationResponse;
    }

    @Override
    public Order submitOrder(String accessToken, Order order) {
        final SubmitOrderRequest submitOrderRequest = new SubmitOrderRequest();
        submitOrderRequest.accessToken = accessToken;
        submitOrderRequest.order = order;

        final OrderConfirmation submitOrderResponse = request(
                submitOrderRequest, new TypeReference<Response<OrderConfirmation>>() {});

        return submitOrderResponse.order;
    }

    @Override
    public Order retrieveOrderAsOwner(String orderId, String ownerToken) {
        final GetOrderRequest getOrderRequest = new GetOrderRequest();
        getOrderRequest.orderId = orderId;
        getOrderRequest.viewMode = Actors.customer;
        getOrderRequest.ownerToken = ownerToken;

        final Order getOrderResponse = request(
                getOrderRequest, new TypeReference<Response<Order>>() {});

        return getOrderResponse;
    }

    @Override
    public Order retrieveOrderAsRestaurant(String accessToken, String orderId) {
        final GetOrderRequest getOrderRequest = new GetOrderRequest();
        getOrderRequest.accessToken = accessToken;
        getOrderRequest.orderId = orderId;
        getOrderRequest.viewMode = Actors.restaurant;

        final Order getOrderResponse = request(
                getOrderRequest, new TypeReference<Response<Order>>() {});

        return getOrderResponse;
    }


    @Override
    public List<SearchResult> search(Filter filter, int limit) {
        final SearchRequest searchRequest = new SearchRequest();
        searchRequest.filter = filter;
        searchRequest.limit = limit;

        final SearchResponse searchResponse = request(
                searchRequest, new TypeReference<Response<SearchResponse>>() {});

        return searchResponse.results;
    }

    @Override
    public void changeOrganizationLocale(String accessToken, String organizationId, Locale locale) {
        final ChangeOrganizationLocaleRequest changeOrganizationLocaleRequest = new ChangeOrganizationLocaleRequest();
        changeOrganizationLocaleRequest.accessToken = accessToken;
        changeOrganizationLocaleRequest.organizationId = organizationId;
        changeOrganizationLocaleRequest.locale = locale;

        request(changeOrganizationLocaleRequest, new TypeReference<Response<Object>>() {});
    }

    @Override
    public Role getRole(String accessToken, String organizationId) {
        final GetRoleRequest getRoleRequest = new GetRoleRequest();
        getRoleRequest.accessToken = accessToken;
        getRoleRequest.organizationId = organizationId;

        final Role getRoleResponse = request(
                getRoleRequest, new TypeReference<Response<Role>>() {});

        return getRoleResponse;
    }

    @Override
    public List<Order> retrieveNewOrders(String accessToken, String restaurantId) {
        final QueryOrdersRequest queryOrdersRequest = new QueryOrdersRequest();
        queryOrdersRequest.accessToken = accessToken;
        queryOrdersRequest.restaurantIds = Collections.singleton(restaurantId);
        queryOrdersRequest.viewMode = Actors.restaurant;
        queryOrdersRequest.status = Statuses.new_;
        queryOrdersRequest.ordering = "asc";
        queryOrdersRequest.limit = Integer.MAX_VALUE;

        final OrdersResponse queryOrdersResponse = request(
                queryOrdersRequest, new TypeReference<Response<OrdersResponse>>() {});

        return queryOrdersResponse.results;
    }

    @Override
    public List<Order> retrieveOrdersByPhone(String accessToken, String organizationId, String phone, Date modifiedSince, Integer limit) {
        final QueryCustomerOrdersRequest queryCustomerOrdersRequest = new QueryCustomerOrdersRequest();
        queryCustomerOrdersRequest.accessToken = accessToken;
        queryCustomerOrdersRequest.organizationId = organizationId;
        queryCustomerOrdersRequest.customerId = customerByPhone(phone);
        queryCustomerOrdersRequest.modifiedSince = modifiedSince;
        queryCustomerOrdersRequest.limit = limit;

        final OrdersResponse queryCustomerOrdersResponse = request(
                queryCustomerOrdersRequest, new TypeReference<Response<OrdersResponse>>() {});

        return queryCustomerOrdersResponse.results;
    }

    @Override
    public List<Order> retrieveOrdersByEmail(String accessToken, String organizationId, String email, Date modifiedSince, Integer limit) {
        final QueryCustomerOrdersRequest queryCustomerOrdersRequest = new QueryCustomerOrdersRequest();
        queryCustomerOrdersRequest.accessToken = accessToken;
        queryCustomerOrdersRequest.organizationId = organizationId;
        queryCustomerOrdersRequest.customerId = customerByEmail(email);
        queryCustomerOrdersRequest.modifiedSince = modifiedSince;
        queryCustomerOrdersRequest.limit = limit;

        final OrdersResponse queryCustomerOrdersResponse = request(
                queryCustomerOrdersRequest, new TypeReference<Response<OrdersResponse>>() {});

        return queryCustomerOrdersResponse.results;
    }

    @Override
    public Order acceptOrder(String accessToken, String orderId, Map<String, String> externalIds) {
        final SetOrderStatusRequest setOrderStatusRequest = new SetOrderStatusRequest();
        setOrderStatusRequest.accessToken = accessToken;
        setOrderStatusRequest.orderId = orderId;
        setOrderStatusRequest.status = Statuses.accepted;
        setOrderStatusRequest.externalIds = externalIds;

        final Order setOrderStatusResponse = request(
                setOrderStatusRequest, new TypeReference<Response<Order>>() {});

        return setOrderStatusResponse;
    }

    @Override
    public Order rejectOrder(String accessToken, String orderId, String comment) {
        final SetOrderStatusRequest setOrderStatusRequest = new SetOrderStatusRequest();
        setOrderStatusRequest.accessToken = accessToken;
        setOrderStatusRequest.orderId = orderId;
        setOrderStatusRequest.status = Statuses.canceled;
        setOrderStatusRequest.comment = comment;

        final Order setOrderStatusResponse = request(
                setOrderStatusRequest, new TypeReference<Response<Order>>() {});

        return setOrderStatusResponse;
    }

    @Override
    public Reservation submitReservation(String accessToken, Reservation reservation) {
        throw new RestaurantsException("deprecated, use API v2");
    }

    @Override
    public Reservation retrieveReservationAsOwner(String ownerToken, String reservationId) {
        throw new RestaurantsException("deprecated, use API v2");
    }

    @Override
    public Reservation retrieveReservationAsRestaurant(String accessToken, String reservationId) {
        throw new RestaurantsException("deprecated, use API v2");
    }

    @Override
    public List<Reservation> retrieveUnhandledReservations(String accessToken, String restaurantId) {
        final QueryUnhandledReservationsRequest queryUnhandledReservationsRequest = new QueryUnhandledReservationsRequest();
        queryUnhandledReservationsRequest.accessToken = accessToken;
        queryUnhandledReservationsRequest.organizationId = restaurantId;
        queryUnhandledReservationsRequest.viewMode = Actors.restaurant;

        final ReservationsResponse queryUnhandledReservationsResponse = request(
                queryUnhandledReservationsRequest, new TypeReference<Response<ReservationsResponse>>() {});

        return queryUnhandledReservationsResponse.results;
    }

    @Override
    public Reservation setReservationStatusAsRestaurant(String accessToken, String reservationId, String status, String comment) {
        final SetReservationStatusRequest setReservationStatusRequest = new SetReservationStatusRequest();
        setReservationStatusRequest.accessToken = accessToken;
        setReservationStatusRequest.reservationId = reservationId;
        setReservationStatusRequest.status = status;
        setReservationStatusRequest.actingAs = Actors.restaurant;
        setReservationStatusRequest.comment = comment;

        final Reservation setReservationStatusResponse = request(
                setReservationStatusRequest, new TypeReference<Response<Reservation>>() {});

        return setReservationStatusResponse;
    }

    @Override
    public Reservation setReservationStatusAsOwner(String ownerToken, String reservationId, String status, String comment) {
        final SetReservationStatusRequest setReservationStatusRequest = new SetReservationStatusRequest();
        setReservationStatusRequest.ownerToken = ownerToken;
        setReservationStatusRequest.reservationId = reservationId;
        setReservationStatusRequest.status = status;
        setReservationStatusRequest.actingAs = Actors.customer;
        setReservationStatusRequest.comment = comment;

        final Reservation setReservationStatusResponse = request(
                setReservationStatusRequest, new TypeReference<Response<Reservation>>() {});

        return setReservationStatusResponse;
    }

    @Override
    public List<Reservation> retrieveReservationsByPhone(String accessToken, String organizationId, String phone, Date modifiedSince, Integer limit) {
        final QueryCustomerReservationsRequest queryCustomerReservationsRequest = new QueryCustomerReservationsRequest();
        queryCustomerReservationsRequest.accessToken = accessToken;
        queryCustomerReservationsRequest.organizationId = organizationId;
        queryCustomerReservationsRequest.customerId = customerByPhone(phone);
        queryCustomerReservationsRequest.modifiedSince = modifiedSince;
        queryCustomerReservationsRequest.limit = limit;

        final ReservationsResponse queryCustomerReservationsResponse = request(
                queryCustomerReservationsRequest, new TypeReference<Response<ReservationsResponse>>() {});

        return queryCustomerReservationsResponse.results;
    }

    @Override
    public List<Reservation> retrieveReservationsByEmail(String accessToken, String organizationId, String email, Date modifiedSince, Integer limit) {
        final QueryCustomerReservationsRequest queryCustomerReservationsRequest = new QueryCustomerReservationsRequest();
        queryCustomerReservationsRequest.accessToken = accessToken;
        queryCustomerReservationsRequest.organizationId = organizationId;
        queryCustomerReservationsRequest.customerId = customerByEmail(email);
        queryCustomerReservationsRequest.modifiedSince = modifiedSince;
        queryCustomerReservationsRequest.limit = limit;

        final ReservationsResponse queryCustomerReservationsResponse = request(
                queryCustomerReservationsRequest, new TypeReference<Response<ReservationsResponse>>() {});

        return queryCustomerReservationsResponse.results;
    }

    @Override
    public void mapInstance(String accessToken, String instanceId, String organizationId) {
        final SetWixAppMappingRequest setWixAppMappingRequest = new SetWixAppMappingRequest();
        setWixAppMappingRequest.accessToken = accessToken;
        setWixAppMappingRequest.instanceId = instanceId;
        setWixAppMappingRequest.organizationId = organizationId;

        request(setWixAppMappingRequest, new TypeReference<Response<Object>>() {});
    }

    @Override
    public void mapInstanceWithWixInstance(String instance, String instanceId, String organizationId) {
        final SetWixAppMappingRequest setWixAppMappingRequest = new SetWixAppMappingRequest();
        setWixAppMappingRequest.instance = instance;
        setWixAppMappingRequest.instanceId = instanceId;
        setWixAppMappingRequest.organizationId = organizationId;

        request(setWixAppMappingRequest, new TypeReference<Response<Object>>() {});
    }

    @Override
    public Organization retrieveOrganizationForInstance(String instanceId) {
        final GetAppMappedObjectRequest getAppMappedObjectRequest = new GetAppMappedObjectRequest();
        getAppMappedObjectRequest.appId = new AppId();
        getAppMappedObjectRequest.appId.platform = AppId.NS_WIX;
        getAppMappedObjectRequest.appId.id = instanceId;
        getAppMappedObjectRequest.appId.version = "1";
        getAppMappedObjectRequest.full = false;

        final Organization getAppMappedObjectResponse = request(
                getAppMappedObjectRequest, new TypeReference<Response<Organization>>() {});

        return getAppMappedObjectResponse;
    }

    @Override
    public void deleteOrganization(String accessToken, String organizationId) {
        final DeleteOrganizationRequest deleteOrganizationRequest = new DeleteOrganizationRequest();
        deleteOrganizationRequest.accessToken = accessToken;
        deleteOrganizationRequest.organizationId = organizationId;

        request(deleteOrganizationRequest, new TypeReference<Response<Object>>() {});
    }

    @Override
    public void deleteCustomerByPhone(String accessToken, String organizationId, String phone) {
        final DeleteCustomerRequest deleteCustomerRequest = new DeleteCustomerRequest();
        deleteCustomerRequest.accessToken = accessToken;
        deleteCustomerRequest.organizationId = organizationId;
        deleteCustomerRequest.customerId = customerByPhone(phone);

        request(deleteCustomerRequest, new TypeReference<Response<Object>>() {});
    }

    @Override
    public void deleteCustomerByEmail(String accessToken, String organizationId, String email) {
        final DeleteCustomerRequest deleteCustomerRequest = new DeleteCustomerRequest();
        deleteCustomerRequest.accessToken = accessToken;
        deleteCustomerRequest.organizationId = organizationId;
        deleteCustomerRequest.customerId = customerByEmail(email);

        request(deleteCustomerRequest, new TypeReference<Response<Object>>() {});
    }

    private <T> T request(Request request, TypeReference<Response<T>> responseType) {
        try {
            return openrest.request(request, responseType);
        } catch (IOException e) {
            throw new CommunicationException(e.getMessage(), e);
        } catch (OpenrestException e) {
            throw translateException(e);
        }
    }

    private static RestaurantsException translateException(OpenrestException e) {
        switch (e.error()) {
            case Error.ERROR_NO_PERMISSION:
                return new NoPermissionException(e.errorMessage(), e);
            case Error.ERROR_INVALID_DATA:
                return new InvalidDataException(e.errorMessage(), e);
            case Error.ERROR_INTERNAL:
                return new InternalException(e.errorMessage(), e);
            case Error.ERROR_NOT_FOUND:
                return new NotFoundException(e.errorMessage(), e);
            default:
                return new RestaurantsException(e.error() + "|" + e.errorMessage(), e);
        }
    }

    private static ClientId customerByPhone(String phone) {
        return new ClientId(ClientNamespaces.phone, phone, null, false);
    }

    private static ClientId customerByEmail(String email) {
        return new ClientId(ClientNamespaces.email, email, null, false);
    }
}

File: wix-restaurants-java-client/src/main/java/com/wix/restaurants/DefaultWixRestaurantsClient.scala
package com.wix.restaurants

import java.net.URLEncoder
import java.util.{Date, List => JList, Map => JMap}

import akka.actor.ActorSystem
import akka.http.javadsl.model.headers.Authorization
import akka.http.scaladsl.client.RequestBuilding.{Delete, Get, Post, Put}
import com.openrest.v1_1._
import com.wix.pay.smaug.client.model.CreditCardToken
import com.wix.rest.rfc7807.api.model.ErrorResponse
import com.wix.rest.rfc7807.client.AkkaRestClient
import com.wix.restaurants.authentication.model.{Namespaces, User => AuthenticationUser}
import com.wix.restaurants.authentication.{DefaultWixRestaurantsAuthenticationClient, WixRestaurantsAuthenticationClient}
import com.wix.restaurants.authorization.{AuthorizationClient, DefaultAuthorizationClient}
import com.wix.restaurants.exceptions._
import com.wix.restaurants.i18n.Locale
import com.wix.restaurants.json.Json
import com.wix.restaurants.orders.{Orders, Statuses => OrderStatuses}
import com.wix.restaurants.reservations.{Reservation, Reservations, Statuses => ReservationStatuses}

import scala.collection.JavaConverters._
import scala.concurrent.duration._
import scala.concurrent.{Await, ExecutionContext}

class DefaultWixRestaurantsClient(apiUrl: String = "https://api.wixrestaurants.com/v2",
                                  authApiUrl: String = "https://auth.wixrestaurants.com/v2",
                                  readTimeout: Duration = Duration.Inf) extends WixRestaurantsClient {

  private val authenticationClient: WixRestaurantsAuthenticationClient = new DefaultWixRestaurantsAuthenticationClient(
    authApiUrl, Some(readTimeout))
  private val authorizationClient: AuthorizationClient = new DefaultAuthorizationClient(
    apiUrl = apiUrl)

  private implicit val system: ActorSystem = ActorSystem("akka-wix-restaurants-client-system")
  private implicit val executionContext: ExecutionContext = system.dispatcher
  private val client: AkkaRestClient = new AkkaRestClient(errorResponseAsException = ExceptionTranslator.asException)


  override def getAuthenticationClient: WixRestaurantsAuthenticationClient = authenticationClient

  override def getAuthorizationClient: AuthorizationClient = authorizationClient

  override def retrieveRestaurantInfo(restaurantId: String): RestaurantFullInfo = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/full")
    Await.result(client.execute(request) withResult[RestaurantFullInfo](), readTimeout)
  }

  override def retrieveOrganization(organizationId: String): Organization = {
    val request = Get(s"$apiUrl/organizations/$organizationId")
    Await.result(client.execute(request) withResult[Organization](), readTimeout)
  }

  override def setOrganization(accessToken: String, organization: Organization): Organization = {
    val request = Put(s"$apiUrl/organizations/${organization.id}", Json.stringify(organization))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Organization](), readTimeout)
  }

  override def changeOrganizationLocale(accessToken: String, organizationId: String, locale: Locale): Unit = {
    val request = Post(s"$apiUrl/organizations/$organizationId/change_locale?locale=$locale")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withoutResult(), readTimeout)
  }

  override def getNotifications(accessToken: String , organizationId: String): Notifications = {
    val request = Get(s"$apiUrl/organizations/$organizationId/notifications")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Notifications](), readTimeout)
  }

  override def setNotifications(accessToken: String, organizationId: String, notifications: Notifications): Notifications = {
    val request = Put(s"$apiUrl/organizations/$organizationId/notifications", Json.stringify(notifications))
        .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Notifications](), readTimeout)
  }

  override def getSecrets(accessToken: String , organizationId: String): Secrets = {
    val request = Get(s"$apiUrl/organizations/$organizationId/secrets")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Secrets](), readTimeout)
  }

  override def setSecrets(accessToken: String, organizationId: String, secrets: Secrets): Secrets = {
    val request = Put(s"$apiUrl/organizations/$organizationId/secrets", Json.stringify(secrets))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Secrets](), readTimeout)
  }

  override def getMenu(accessToken: String , restaurantId: String): Menu = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/menu")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Menu](), readTimeout)
  }

  override def setMenu(accessToken: String, restaurantId: String, menu: Menu): Menu = {
    val request = Put(s"$apiUrl/organizations/$restaurantId/menu", Json.stringify(menu))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Menu](), readTimeout)
  }

  override def submitOrder(accessToken: String, order: Order): Order = {
    val anonymousRequest = Post(s"$apiUrl/organizations/${order.restaurantId}/orders", Json.stringify(order))
    val request = Option(accessToken) match {
      case Some(at) => anonymousRequest.addHeader(Authorization.oauth2(at))
      case None => anonymousRequest
    }
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def retrieveOrderAsRestaurant(accessToken: String, restaurantId: String, orderId: String): Order = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/orders/$orderId?viewMode=${Actors.restaurant}")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def retrieveOrderAsOwner(accessToken: String, restaurantId: String, orderId: String): Order = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/orders/$orderId?viewMode=${Actors.customer}")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def retrieveNewOrders(accessToken: String, restaurantId: String): JList[Order] = {
    retrieveOrdersAsRestaurant(accessToken, restaurantId, OrderStatuses.new_, 10000)
  }

  override def retrieveOrdersAsRestaurant(accessToken: String, restaurantId: String, status: String, limit: Int): JList[Order] = {
    val statusPart = Option(status).map { theStatus => s"&status=$theStatus" }.getOrElse("")
    val request = Get(s"$apiUrl/organizations/$restaurantId/orders?viewMode=${Actors.restaurant}$statusPart&limit=$limit")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result[Orders](client.execute(request) withResult[Orders](), readTimeout).results
  }

  override def retrieveOrdersByPhone(accessToken: String, organizationId: String, phone: String, modifiedSince: Date, limit: Integer): JList[Order] = {
    retrieveUserOrders(accessToken, organizationId, new AuthenticationUser(Namespaces.phone, phone), modifiedSince, limit)
  }

  override def retrieveOrdersByEmail(accessToken: String, organizationId: String, email: String, modifiedSince: Date, limit: Integer): JList[Order] = {
    retrieveUserOrders(accessToken, organizationId, new AuthenticationUser(Namespaces.email, email), modifiedSince, limit)
  }

  private def retrieveUserOrders(accessToken: String, organizationId: String, user: AuthenticationUser, modifiedSince: Date, limit: Integer): JList[Order] = {
    val modifiedSinceTimestamp = Option(modifiedSince).map { _.getTime }.getOrElse(0L)
    val actualLimit = Option(limit).map { _.toInt }.getOrElse(1000000)
    val request = Get(s"$apiUrl/organizations/$organizationId/orders?viewMode=${Actors.restaurant}&user=${user.ns}:${URLEncoder.encode(user.id, "UTF-8")}&modified=gte:$modifiedSinceTimestamp&limit=$actualLimit")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result[Orders](client.execute(request) withResult[Orders](), readTimeout).results
  }

  override def acceptOrder(accessToken: String, restaurantId: String, orderId: String, externalIds: JMap[String, String]): Order = {
    val request = Post(s"$apiUrl/organizations/$restaurantId/orders/$orderId/accept?as=${Actors.restaurant}", Json.stringify(Comment(None)))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def rejectOrder(accessToken: String, restaurantId: String, orderId: String, comment: String): Order = {
    val request = Post(s"$apiUrl/organizations/$restaurantId/orders/$orderId/cancel?as=${Actors.restaurant}", Json.stringify(Comment(Option(comment))))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def setOrderProperties(accessToken: String, restaurantId: String, orderId: String, properties: JMap[String, String]): Order = {
    val request = Put(s"$apiUrl/organizations/$restaurantId/orders/$orderId/properties", Json.stringify(properties))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def confirmOrderCashier(organizationId: String, orderId: String): Order = {
    val request = Post(s"$apiUrl/organizations/$organizationId/orders/$orderId/confirmCashier?as=${Actors.restaurant}")
    Await.result(client.execute(request) withResult[Order](), readTimeout)
  }

  override def submitReservation(accessToken: String, reservation: Reservation): Reservation = {
    val request = Post(s"$apiUrl/organizations/${reservation.restaurantId}/reservations", Json.stringify(reservation))
    Option(accessToken).foreach { theAccessToken => request.addHeader(Authorization.oauth2(theAccessToken)) }
    Await.result(client.execute(request) withResult[Reservation](), readTimeout)
  }

  override def retrieveReservationAsOwner(accessToken: String, restaurantId: String, reservationId: String): Reservation = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/reservations/$reservationId?viewMode=${Actors.customer}")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Reservation](), readTimeout)
  }

  override def retrieveReservationAsRestaurant(accessToken: String, restaurantId: String, reservationId: String): Reservation = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/reservations/$reservationId?viewMode=${Actors.restaurant}")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[Reservation](), readTimeout)
  }

  override def retrieveUnhandledReservations(accessToken: String, restaurantId: String): JList[Reservation] = {
    val request = Get(s"$apiUrl/organizations/$restaurantId/reservations?viewMode=${Actors.restaurant}&unhandled=true")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result[Reservations](client.execute(request) withResult[Reservations](), readTimeout).results
  }

  override def setReservationStatusAsRestaurant(accessToken: String, restaurantId: String, reservationId: String, status: String, comment: String): Reservation = {
    status match {
      case ReservationStatuses.accepted =>
        val request = Post(s"$apiUrl/organizations/$restaurantId/reservations/$reservationId/accept?as=${Actors.restaurant}",
          Json.stringify(Comment(comment = Option(comment))))
          .addHeader(Authorization.oauth2(accessToken))
        Await.result(client.execute(request) withResult[Reservation](), readTimeout)

      case ReservationStatuses.canceled =>
        val request = Post(s"$apiUrl/organizations/$restaurantId/reservations/$reservationId/cancel?as=${Actors.restaurant}",
          Json.stringify(Comment(comment = Option(comment))))
          .addHeader(Authorization.oauth2(accessToken))
        Await.result(client.execute(request) withResult[Reservation](), readTimeout)
    }
  }

  override def setReservationStatusAsOwner(accessToken: String, restaurantId: String, reservationId: String, status: String, comment: String): Reservation = {
    status match {
      case ReservationStatuses.canceled =>
        val request = Post(s"$apiUrl/organizations/$restaurantId/reservations/$reservationId/cancel?as=${Actors.customer}",
          Json.stringify(Comment(comment = Option(comment))))
          .addHeader(Authorization.oauth2(accessToken))
        Await.result(client.execute(request) withResult[Reservation](), readTimeout)
    }
  }

  override def retrieveReservationsByPhone(accessToken: String, organizationId: String, phone: String, modifiedSince: Date, limit: Integer): JList[Reservation] = {
    retrieveUserReservations(accessToken, organizationId, new AuthenticationUser(Namespaces.phone, phone), modifiedSince, limit)
  }

  override def retrieveReservationsByEmail(accessToken: String, organizationId: String, email: String, modifiedSince: Date, limit: Integer): JList[Reservation] = {
    retrieveUserReservations(accessToken, organizationId, new AuthenticationUser(Namespaces.email, email), modifiedSince, limit)
  }

  private def retrieveUserReservations(accessToken: String, organizationId: String, user: AuthenticationUser, modifiedSince: Date, limit: Integer): JList[Reservation] = {
    val modifiedSinceTimestamp = Option(modifiedSince).map { _.getTime }.getOrElse(0L)
    val actualLimit = Option(limit).map { _.toInt }.getOrElse(1000000)
    val request = Get(s"$apiUrl/organizations/$organizationId/reservations?viewMode=${Actors.restaurant}&user=${user.ns}:${URLEncoder.encode(user.id, "UTF-8")}&modified=gte:$modifiedSinceTimestamp&limit=$actualLimit")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result[Reservations](client.execute(request) withResult[Reservations](), readTimeout).results
  }

  override def deleteOrganization(accessToken: String, organizationId: String): Unit = {
    val request = Delete(s"$apiUrl/organizations/$organizationId")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withoutResult(), readTimeout)
  }

  override def deleteCustomerByPhone(accessToken: String, organizationId: String, phone: String): Unit = {
    deleteCustomer(accessToken, organizationId, new AuthenticationUser(Namespaces.phone, phone))
  }

  override def deleteCustomerByEmail(accessToken: String, organizationId: String, email: String): Unit = {
    deleteCustomer(accessToken, organizationId, new AuthenticationUser(Namespaces.email, email))
  }

  override def getMyAccount(accessToken: String): ClientInfo = {
    val request = Get(s"$apiUrl/me/account")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[ClientInfo](), readTimeout)
  }

  override def setMyAccount(accessToken: String, account: ClientInfo): ClientInfo = {
    val request = Put(s"$apiUrl/me/account", Json.stringify(account))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[ClientInfo](), readTimeout)
  }

  override def addMyAccountCards(accessToken: String, cardTokens: JList[CreditCardToken]): ClientInfo = {
    val request = Post(s"$apiUrl/me/account/cards", Json.stringify(CardTokens(cardTokens.asScala)))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[ClientInfo](), readTimeout)
  }

  private def deleteCustomer(accessToken: String, organizationId: String, customer: AuthenticationUser): Unit = {
    val request = Post(s"$apiUrl/organizations/$organizationId/delete_customer", Json.stringify(customer))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withoutResult(), readTimeout)
  }
}


private object ExceptionTranslator {
  def asException(errorResponse: ErrorResponse): RuntimeException = {
    errorResponse.`type` match {
      case Errors.InvalidData => new InvalidDataException(errorResponse.detail.orNull)
      case Errors.Forbidden => new NoPermissionException(errorResponse.detail.orNull)
      case Errors.Conflict => new ConflictException(errorResponse.detail.orNull)
      case Errors.NotFound => new NotFoundException(errorResponse.detail.orNull)
      case Errors.Internal => new InternalException(errorResponse.detail.orNull)
      case Errors.TemporarilyUnavailable => new TemporarilyUnavailableException(errorResponse.detail.orNull)
      case _ => new RestaurantsException(s"Type: '${errorResponse.`type`}'${errorResponse.detail.map(detail => s", Detail: '$detail'").getOrElse("")}")
    }
  }
}
